File: 538_Convert_BST_to_Greater_Tree.java
/*
https://leetcode.com/problems/convert-bst-to-greater-tree/
Given a Binary Search Tree (BST), convert it to a Greater Tree such that every key of the original BST is changed to
the original key plus sum of all keys greater than the original key in BST.

Reverse inorder traversal. Add every node.val to running sum and update node.val with running sum.
Time - O(n)
Space - O(1)
*/

class Solution {

    int sum = 0;

    public TreeNode convertBST(TreeNode root) {
        reverseInorder(root);
        return root;
    }

    private void reverseInorder(TreeNode node) {
        if (node == null)
            return;
        reverseInorder(node.right);
        sum += node.val;
        node.val = sum;
        reverseInorder(node.left);
    }
}

File: python_1_to_1000/538_Convert_BST_to_Greater_Tree.py
_author_ = 'jake'
_project_ = 'leetcode'

# https://leetcode.com/problems/convert-bst-to-greater-tree/
# Given a Binary Search Tree (BST), convert it to a Greater Tree such that every key of the original BST is changed to
# the original key plus sum of all keys greater than the original key in BST.

# Reverse inorder traversal. Visit right subtree (containing all greater values), update node value and running sum,
# then visit left subtree.
# Time - O(n)
# Space - O(1)

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        self.running_sum = 0

        def inorder(node):
            if not node:
                return

            inorder(node.right)

            node.val += self.running_sum
            self.running_sum = node.val

            inorder(node.left)

        inorder(root)
        return root